### ISSUE TITLE
opt: look at using constraints to plan parameterized spans for simplified exec time lookup joins

### ISSUE THREAD

**Issue by cucaroach** [cockroachdb/cockroach#65773]
@rytaft and I discussed pulling the expr walking code out of the lookup join for the multi span stuff.    I had a vague notion that a compile time approach using parameterized spans would be simplify the exec time stuff and leave less room for error but come to find out @rytaft actually went down the path and ran into some issues.   We want to get there eventually, this is a TODO item issue to record what issues we ran into and how we're gonna get around them.

**Comment by rytaft** [comment]
Thanks for opening the issue, @cucaroach!

In addition to simplifying the execution code for lookup joins, addressing this issue could also support expansion of the [placeholder fast path](https://github.com/cockroachdb/cockroach/issues/57223).

The idea I was investigating before was to allow using a `tree.Placeholder` in a `constraint.Key`. This works in theory since `tree.Placeholder` implements the `tree.Datum` interface, but it causes issues when the constraint library tries to compare keys for the purpose of merging spans, etc. Almost all implementations of the `tree.Datum` functions for `tree.Placeholder` call `mustGetValue`, which returns an error when no value is found for the placeholder.

To make this work, we would need a way to disable certain types of operations in the constraint library when placeholders are present. However, we'd still want to perform the usual operations on the non-placeholder keys. For example, `[/$1 - /3] (/3 - /5]` should still be simplified to `[/$1 - /5]`. I started implementing the necessary changes to make this work, but it turned out to be a very deep rabbit hole requiring a lot of changes to both the `constraint` and `idxconstraint` libraries.

It's possible that with more effort this approach could work, but it's also equally possible that it would be better to create a separate library specifically for constraints with one or more non-constant keys. It's also possible that the current approach of passing a `tree.TypedExpr` around, although unsatisfying, is the cleanest approach. More investigation is needed to determine which option is best.

**Comment by cucaroach** [comment]
@rytaft and @yuzefovich I'm guessing this might help with doing a full vectorized version of the join reader, I wonder if I should take a crack at it in Sept milestone?

**Comment by rytaft** [comment]
Can you give more details about how it would help with building a vectorized version of join reader? It seems to me like these issues are orthogonal.

**Comment by cucaroach** [comment]
I was assuming the vectorized join reader is going to have to re-implement all the lookupExpr stuff, maybe that's wrong?

**Comment by yuzefovich** [comment]
To be honest, I haven't spent much time thinking about how we would vectorize the lookup join. If we take a similar approach to what Drew had for the index join, on a quick glance it does look like we'll need to figure out how to generate spans based on the `lookupExpr` (i.e. vectorize the logic of `multiSpanGenerator`).

**Comment by rytaft** [comment]
True, but the original expression is only used once when the `multiSpanGenerator` is initialized, in `fillInIndexColInfos`. I would guess that most of the logic in `fillInIndexColInfos` can probably just be copied.

It seems more likely that `generateSpans` will need to change, and that does not use the original `lookupExpr`.
